daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ExpressionEvaluatingUnparsers.scala

```diff
@@ -77,30 +76,52 @@ final class SetVariableUnparser(
 
 }
 
+final class NewVariableInstanceSuspendableExpression(
+  override val expr: CompiledExpression[AnyRef],
+  override val rd: VariableRuntimeData)
+  extends SuspendableExpression {
+
+  override protected def processExpressionResult(ustate: UState, v: DataValuePrimitive): Unit = {
+    ustate.variableMap.newVariableInstance(rd, v)
+  }
+
+  override protected def maybeKnownLengthInBits(ustate: UState) = MaybeULong(0)
+}
+
```

Comment:
Hm, I think you might be right.  What probably needs to be done is to add an additional variable state, VariableSuspended, which is then changed once the expression evaluates.  Attempting to read or set a variable that is suspended would then also need to trigger a suspension.

Reply:
One thing to that jumped to my mind, does a setVariable even need to suspend if the defaultValue is suspending (i.e. the vairables is in the new VariableSuspended state)? In this case, we'll never use the defaultValue, so why wait for it to become evaluated if setVariable doesn't need to suspend?

I think the answer here is the defaultValue could access other variables which changes the state of those variables to read, so it probably is important to wait for the defaultValue to evaluate, even if it's value will be overwriten by a setVariable. 

Another unrelated thought about how this all needs to work, say we had something like this:
* NVI variable1, defaultValue = 1
* NVI variable2, defaultValue = "{ /some/future/path/that/requires/suspension + $variable1 }"
* setVariable, variable1 = 2
* element A dfdl:inputValueCalc="{ $variable1 }"

So we create NVI variable1 with an immediate value of defaultValue1. All good.

Then we create a NVI variable2, but the defaultValue must suspend because it accesses some future path. In suspending, it clones the UState (including the current "1" value of variable1). Note that it hasn't read variable1 until the path comes into existance.

We continue the unparse and evaluate the setVariable and change variable1 to a value of 2. This is fine because variable1 hasn't been read yet, because NVI variable2 defaultValue suspended.

Continuing the parse, we evaluate the inputValueCalc and set element A to variable1, which now has a value of 2.

At some point we unsuspend the defaultValue, and set defalutValue to 1. I *think* this works because this suspension is working on a cloned UState. But this uses the 1 value from variable1 because it was cloned. This feels wrong?

And we're done, and it's not clear what the right behavior was.

Note that if variable2 didn' need to suspend (say that long path existed), then variable1 is immedately read when setting variable2's defaultValue, and so the setVariable for variable1 should be an SDE.

So it seems we could get two different behaviors depending if defaultValue suspends or not, and it's not immediately clear to me what the right behavior is. Hopefully I'm missing something?

Reply:
The right behavior  is that suppose a DFDL implementation did not do streaming unparsing at all. There would be no-suspending when reaching forward for element values. 

The behavior one gets in that situation is the correct answer.  Streaming unparsing has to achieve this same behavior, or it is a bug. 

The correct behavior in the above example in steve lawrence's comment  is a runtime SDE due to setVariable variable1 occuring after the variable1 has already been read for its default value. 
 
In Daffodil's streaming implementation, the forward referencing expression causing a suspension should make a copy/clone of all variable state, so such time as it does read variable1, it will get the default value of 1. 

So far so good.

As unparsing continues, the setVariable statement will be encountered, and it will set variable1's value to 2.  It does not know that the expression has suspended but will eventually read variable1.  This is the bug. 

A new state on variable1 indicating it is pending a read by a suspension won't help. This doesn't work in general, because expressions can have if-then-else logic such that some conditional branches evaluate certain variables, not others. So until we evaluate an expression we can't know what variables it would read. 

I believe we have a situation close to a classic compiler anti-dependence here. 

Fact is, this setVariable should be itself suspended until after the variable-read in the expression for variable2. 

The current daffodil expression implementation won't even attempt to read variable1's value until the suspension completes; hence, variable1 wouldn't even get marked as 'pending read' state until the forward reference has been resolved.

Reply:
Doing the full dependency analysis of every read/write of a variable across the DFDL schema is one way to solve this, albeit a complex one.

I think a "workaround" fix might be this: in variable direction both or unparseOnly, disallow setVariable after a NVI with a default value. This requires the user to either remove the default value from the NVI, or remove the setVariable. This requires analysis of whether there are setVariable statements in scope of an unparse-direction NVI having default value, and that's itself somewhat tricky, but it is far easier than a full dependency analysis.

Reply:
An easier check could be if a given variable is used anywhere in the schema by a NVI statement having default value, and the variable has unparse-direction, that no setVariable is allowed anywhere for that variable. This eliminates the need for a precise roll-up of which setVariable statements are in scope of a particular NVI or not. 

Such a variable is what we might call "bindable only" or non-settable. I think this fixes many things, and DFDL schemas that actually try to use setVariable with such variables can split a variable into two variables. Use one for binding via NVI with defaults, and the other for the setVariable purposes, it's NVIs would not have any default values.

Reply:
As this is a dfdlx experimental feature, I think we need a wiki page description of it, and that would have a section detailing these implementation-specific restrictions on how to use variables at unparse time due to the streaming unparser behavior.

Reply:
Lastly this scenario SL came up with needs a test that either shows the problem, or shows it not occurring.

Reply:
Thinking more on this, I think there's maybe some other complexities related to variables, and doesn't even include NVI.

Imagine we having a variable with a constant static default, and then a setVariable to set that variable but that must suspend due to a forward lookahead. Now if an an element tries to read that variable, it must also suspend until that setVariable completes. Something to consider with this is that setVariable suspension is working on a cloned ustate. So the cloned ustate must do a shallow copy of the variable map so that when it does modify the variable map, other suspensions will see that change. I think this may be how it works, but worth having some tests to consider.

But also note that since this is a shallow copy, it means other suspensions/parsers could modify this variable map, and could maybe potentially affect the suspension once it unsuspends? Maybe something like this:

* variable1 defaultValue = 1
* variable2 defaultValue = 2
* setVariable variable1, value = /forward/path/forcing/suspension + variable2
* setVariable variable2, value = 3

So setVariable variable1 must suspend. Then setVariable varible2 changes the value of variable2 to 3. Then at some point when setVariable variable1 unsuspends, it gets the value of 3 for variable2 (assuming shallow variable map copy in the ustate clone). The correct result is setVariable2 should cause an SDE because setVariable variable1 should have read variable2. So this too has this order dependency stuff going on and doesn't involve NVI. I think this is basically the same thing as my previous scenario, but doesn't use NVI.

Reply:
This may ultimately just be a flaw in the DFDL language notion of variables. They're supposed to be single-assignment variables, but this mixture of defaultValue and setVariable where a read locks down the default value was always a sketchy design point because things become very order dependent, and since the unparser does things out of order due to suspensions and forward reference, the tenuous interactions of these things become problematic. 

A way to fix this is to say that for any variable that has unparse direction (so both or unparseOnly) you have to pick whether you use *only* defaultValue expressions to bind values to it, or whether you will use setVariable statements, in which case the variable cannot have any default value in the base defineVariable nor in any newVariableInstance statements.  This I think fixes all these issues. A variable is either defaultable, and not settable, or it is settable and not defaultable.  I think these restrictions are very livable.